Improve `get_block_templates` performance

oandregal opened the pull request:
Trac ticket https://core.trac.wordpress.org/ticket/57756

## What

This PR improves the performance of block template resolution.

## Why

It provides a performance improvement for block themes. This change should not affect themes that don't have block templates, given they [bail early](https://github.com/WordPress/wordpress-develop/blob/4049c5d3543e59f92d48fd9cef735420996dce2e/src/wp-includes/block-template.php#L51).

The effect of this PR according to the [performance CI job](https://github.com/WordPress/wordpress-develop/actions/runs/4733710372/jobs/8401618173?pr=4097):

- Home classic theme

| Metric (ms) | This PR | Base | Improvement |
| --- | --- | --- | --- |
| wpBeforeTemplate | 17.02 | 19.47 | +12.58% |
|    wpTemplate    | 41.06 |  44.40 | +7.52% |
|     wpTotal      | 57.92 | 64.23 | +9.82% |

- Home block theme

| Metric (ms) | This PR | Base | Improvement % |
| --- | --- | --- | --- |
| wpBeforeTemplate | 23.34 | 24.42 | +4.42% |
|    wpTemplate    | 44.27 | 55.24 | +19.85% |
|     wpTotal      | 67.63 | 79.56 | +14.99% |


## How

The `get_block_templates` function is responsible to find block templates that match a given search. The function is provided a query parameter that is used to find the relevant user templates (database) and theme templates (file directory). The query parameter includes data such the slugs of the templates, the areas of the template parts, etc.

I've found that the `get_block_templates` retrieves and processes [all block templates provided by the theme](https://github.com/WordPress/wordpress-develop/blob/4049c5d3543e59f92d48fd9cef735420996dce2e/src/wp-includes/block-template-utils.php#L977), to [filter out the ones that don't match](https://github.com/WordPress/wordpress-develop/blob/4049c5d3543e59f92d48fd9cef735420996dce2e/src/wp-includes/block-template-utils.php#L997) after. This function can be more performant if it filters out the ones that don't match _before_ processing them, so it only computes the data to be used.

The following profile shows the impact of this modification:

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/583546/219652222-2aca88ad-c4b4-4d6f-9b3f-707df626711e.png) | ![image](https://user-images.githubusercontent.com/583546/219652291-3c1f3246-4ac4-49f8-a087-50949f21b48b.png) |

- The time it takes as measured by XDebug profiler: 55ms (before) vs 27ms (after).
- The number of times the `_build_block_template_result_from_file` is called: 22 (before) vs 5 (after).

## How to test

- Start the bundled environment.
- Set `.env` to production by applying the following patch:

```diff
diff --git a/.env b/.env
index 63a8169f64..9b1d52ac50 100644
--- a/.env
+++ b/.env
@@ -54,10 +54,10 @@ LOCAL_DB_TYPE=mysql
 LOCAL_DB_VERSION=5.7
 
 # The debug settings to add to `wp-config.php`.
-LOCAL_WP_DEBUG=true
-LOCAL_WP_DEBUG_LOG=true
-LOCAL_WP_DEBUG_DISPLAY=true
-LOCAL_SCRIPT_DEBUG=true
+LOCAL_WP_DEBUG=false
+LOCAL_WP_DEBUG_LOG=false
+LOCAL_WP_DEBUG_DISPLAY=false
+LOCAL_SCRIPT_DEBUG=false
 LOCAL_WP_ENVIRONMENT_TYPE=local
 
 # The URL to use when running e2e tests.
```
- Run 100 test of the homepage: `seq 100 | xargs -Iz curl -o /dev/null -H 'Cache-Control: no-cache' -s -w "%{time_starttransfer}\n" http://localhost:8889 | xclip -selection clipboard`.
- Copy the results to a spreadsheet like [this one](https://docs.google.com/spreadsheets/d/1K2u5oOh7rd23HBOz3pPAcAi9P8290E0IV2NZhk_E9i4/edit#gid=0) to measure the changes.

## Notes

I submitted this PR to core instead of Gutenberg because it modifies functions that are no longer there:

- `_get_block_templates_files`: no longer present.
- `get_block_templates`: the corresponding `gutenberg_get_block_templates` is [still present](https://github.com/WordPress/gutenberg/blob/trunk/lib/compat/wordpress-6.1/block-template-utils.php#L53), but it'll be removed once WordPress 6.2 is published. It won't be present in Gutenberg when this PR lands (WordPress 6.3 cycle).

oandregal committed:
Filter theme templates (wp_template & wp_template_part) by slug

Only return the theme templates that match the slugs we are looking for,
instead of returning them all.

mukeshpanchal27 commented on src/wp-includes/block-template-utils.php:
```diff
@@ -295,11 +295,16 @@ function _get_block_template_file( $template_type, $slug ) {
  * @return array Template.
  */
-function _get_block_templates_files( $template_type ) {
+function _get_block_templates_files( $template_type, $slugs = array() ) {
```
Quick review. Add new `$slugs` in docblock.

oandregal replied:
Ah, thanks! There are probably much more of these changes (formatting, etc.) that I'll address when this PR is ready. It is just in the prototype stage to gauge impact, I need to look at other areas (see TODO list in the issue description).

I don't know if there's a way to mark a PR as "in progress" in core. In Gutenberg, we have "draft PRs" and the "in progress" label.

mukeshpanchal27 replied:
Ohh. You should mark PR as "draft" so others can understand that some work is still in progress and no one can review it. I hope that makes sense.

oandregal replied:
Ah! I missed the option when first created the PR, but found it in [the sidebar](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request). Thank you :)

oandregal replied:
Provided the relevant PHPDoc for this change at https://github.com/WordPress/wordpress-develop/pull/4097/commits/d532df39a4cd5de61d2483de4ffaa7189a679824

oandregal committed:
Filter wp_template_part by area

Only return template pats that match the area we are looking for,
instead of returning them all.

oandregal committed:
Filter wp_template by postTypes

Only return the theme templates that match the given postTypes,
instead of returning them all.

oandregal committed:
Filter theme templates (wp_template & wp_template_part) by user

Only return items that don't have a user-defined equivalent,
instead of returning them all.

oandregal commented on src/wp-includes/block-template-utils.php:
```diff
@@ -957,34 +1003,10 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
-			if ( $should_include ) {
-				$query_result[] = $template;
-			}
+			$query_result[] = _build_block_template_result_from_file( $template_file, $template_type );
```
This is the key change of this PR: instead of building the template object _for all items_ and filter them _after_, we filter _before_ this step, so we have to build less items. We can do this because all the data used to filter the items is already available in the step before.

oandregal replied:
The impact of this change can be seen in the XDebug profiler by looking at the number of times the `_build_block_template_result_from_file` function is called:

| Before (22 times) | After (5 times) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/583546/219652222-2aca88ad-c4b4-4d6f-9b3f-707df626711e.png) | ![image](https://user-images.githubusercontent.com/583546/219652291-3c1f3246-4ac4-49f8-a087-50949f21b48b.png) |

oandregal committed:
Add PHPDoc for _get_block_templates_files changes

oandregal commented:
ok, this is now ready to review :) cc @youknowriad @ockham @creativecoder @carlomanf as people who has worked in this area, according to https://github.com/WordPress/wordpress-develop/pull/1267

oandregal commented:
The [perf job](https://github.com/WordPress/wordpress-develop/actions/runs/4384673981/jobs/7676429393) reports the following (copy/pasted here in a single table, otherwise it's difficult to compare the results):

- Home classic theme

| Metric (ms) | This PR | Base | Improvement |
| --- | --- | --- | --- |
| wpBeforeTemplate | 24.32 | 25.33 | +3.98% |
|    wpTemplate    |       50.67       |  56.97 | +11.05% |
|     wpTotal      | 75.28  | 80.32 | +6.25% |

- Home block theme

| Metric (ms) | This PR | Base | Improvement |
| --- | --- | --- | --- |
| wpBeforeTemplate | 28.33  |   31.80  | +10.91% |
|    wpTemplate    | 53.37 | 71.39  | +25.24% |
|     wpTotal      | 81.34 |  102.74  | +20.82% |

spacedmonkey commented on src/wp-includes/block-template-utils.php:
```diff
@@ -957,34 +1014,10 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 	}
```
While we are at it can we change these lines. 


```
	if ( isset( $query['slug__in'] ) ) {
		$wp_query_args['post_name__in'] = $query['slug__in'];
	}
```

To 
```
	if ( isset( $query['slug__in'] ) ) {
		$wp_query_args['post_name__in'] = $query['slug__in'];
               $wp_query_args['posts_per_page'] = count( $query['slug__in'] );
	}
```

This limits the query and improves performance.

oandregal replied:
Done at https://github.com/WordPress/wordpress-develop/pull/4097/commits/8a0f1f958ee348ef29fe8e975203db9a33a1a736

spacedmonkey commented on src/wp-includes/block-template-utils.php:
```diff
@@ -290,19 +290,36 @@ function _get_block_template_file( $template_type, $slug ) {
 	$themes         = array(
-		get_stylesheet() => get_stylesheet_directory(),
-		get_template()   => get_template_directory(),
+		$stylesheet    => get_stylesheet_directory(),
+		get_template() => get_template_directory(),
```
These can be the same value, if it is not a child theme. We could do this. 

```
$stylesheet     = get_stylesheet();
$template        = get_template();

$themes         = array(
            $stylesheet    => get_stylesheet_directory(),
);

if( $stylesheet !== $template ){
   $themes[$template] = get_template_directory();
}
```

This would stop it loop around unnecessarily.

oandregal replied:
Done at https://github.com/WordPress/wordpress-develop/pull/4097/commits/3038e5b27d9ab8dfece28cc940dae0787669565a

youknowriad commented on src/wp-includes/block-template-utils.php:
```diff
@@ -290,19 +290,36 @@ function _get_block_template_file( $template_type, $slug ) {
+ *     Arguments to retrieve templates. Optional, empty by default.
+ *
+ *     @type array  $slug__in  List of slugs to include.
+ *     @type int    $wp_id     Post ID of customized template.
```
This probably doesn't make sense for retrieving files.

jorgefilipecosta replied:
Would it be possible to keep function _get_block_templates_files( $template_type ) as is to keep its meaning and address @youknowriad points but have a not another function called right after with the result of _get_block_templates_files call that receives $query, and $user_templates and filters the templates?

youknowriad replied:
I think query still makes sense here personally, there's no need for two functions but it's the "user templates" filtering that can be done as a post process IMO. Alternatively, we could pass an array of slugs to ignore in the "query".

oandregal replied:
Removed the `wp_id` at https://github.com/WordPress/wordpress-develop/pull/4097/commits/1e76e2c2a47a6bc7cd49c7996d67a8289096ad24 and converted the `user_templates` param to the `slugs__not_in` query param at https://github.com/WordPress/wordpress-develop/pull/4097/commits/581d946f3b261591603953e4bac9072f03900d05

youknowriad commented on src/wp-includes/block-template-utils.php:
```diff
@@ -290,19 +290,36 @@ function _get_block_template_file( $template_type, $slug ) {
+ *     @type string $area      A 'wp_template_part_area' taxonomy value to filter by (for wp_template_part template type only).
+ *     @type string $post_type Post type to get the templates for.
+ * }
+ * @param array $user_templates Array of user templates that match the search. Optional, empty by default.
```
It's really not clear to me why a function that retrieves template from files need template from the database as an argument.

oandregal replied:
addressed at https://github.com/WordPress/wordpress-develop/pull/4097#discussion_r1143256277

youknowriad commented on src/wp-includes/block-template-utils.php:
```diff
@@ -317,6 +334,17 @@ function _get_block_templates_files( $template_type ) {
+			}
+
+			// Skip this item if there's already an user-provided item with the same slug.
+			if ( false !== array_search( $stylesheet . '//' . $template_slug, $user_templates, true ) ) {
```
While I understand the logic, I feel like this breaks the intent of the function (change its meaning), we're not just retrieving file templates anymore.

How much does this contribute to the perf optimization. I'd prefer to remove this check from here personally

oandregal replied:
addressed https://github.com/WordPress/wordpress-develop/pull/4097#discussion_r1143256277

oandregal committed:
Do not query the template directory if it is the same as stylesheet dir.

Co-authored-by: Jonny Harris <spacedmonkey@git.wordpress.org>

oandregal committed:
Remove unused query param

oandregal committed:
Convert the user_templates param into slugs__not_in query param

oandregal committed:
Limit query results by number of slugs

youknowriad commented:
Can't talk about the performance impact but this looks great to me. Thanks for the updates @oandregal

mukeshpanchal27 commented on src/wp-includes/block-template-utils.php:
```diff
@@ -290,20 +290,41 @@ function _get_block_template_file( $template_type, $slug ) {
+ *     @type array  $slug__in      List of slugs to include.
+ *     @type array  $slug__not_in  List of slugs to skip.
+ *     @type string $area          A 'wp_template_part_area' taxonomy value to filter by (for wp_template_part template type only).
+ *     @type string $post_type     Post type to get the templates for.
```
```suggestion
 *     @type array  $slug__in     List of slugs to include.
 *     @type array  $slug__not_in List of slugs to skip.
 *     @type string $area         A 'wp_template_part_area' taxonomy value to filter by (for wp_template_part template type only).
 *     @type string $post_type    Post type to get the templates for.
```

Nit-pick, Remove extra space.

mukeshpanchal27 commented on src/wp-includes/block-template-utils.php:
```diff
@@ -921,8 +982,9 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
-	if ( isset( $query['slug__in'] ) ) {
-		$wp_query_args['post_name__in'] = $query['slug__in'];
+	if ( ! empty( $query['slug__in'] ) ) {
+		$wp_query_args['post_name__in']  = $query['slug__in'];
```
```suggestion
		$wp_query_args['post_name__in']  = array_unique( (array) $query['slug__in'] );
```

Do we needs to check unique array otherwise we get wrong `posts_per_page`. WDYT?

oandregal replied:
Do you mean that if the same slug appears twice, the post will be duplicated? I don't think we need this, no. `post_name__in` becomes [part of the filtering](https://github.com/WordPress/wordpress-develop/blob/36e994b88ceb231b34bde6ae3a4023ceb0f93d4e/src/wp-includes/class-wp-query.php#L2179). I've also tried manually by adding duplicate slugs and it still returned the proper number of posts.

mukeshpanchal27 replied:
In edge case if some how we have duplicate slugs in `$query['slug__in']` value then we get wrong value for the `$wp_query_args['posts_per_page'] = count( $query['slug__in'] );` so `posts_per_page` get wrong value.

Have you gotten my point? Please correct me if I miss anything. Thanks!

oandregal replied:
The code suggestion is attached to the wrong line (`post_name__in` instead `posts_per_page`), that's why it was confusing. I've inspected the file and see that there is a line where we do that to `posts_per_page` so I've committed the suggestion in https://github.com/WordPress/wordpress-develop/pull/4097/commits/ae409e966d5751a577901ca7cc72473ada772ed9. Thanks!

mukeshpanchal27 commented on src/wp-includes/block-template-utils.php:
```diff
@@ -317,6 +338,17 @@ function _get_block_templates_files( $template_type ) {
 			);
+
+			// Skip this item if its slug doesn't match any of the slugs to include.
+			if ( ! empty( $slugs_to_include ) && ( false === array_search( $template_slug, $slugs_to_include, true ) ) ) {
```
```suggestion
			if ( ! empty( $slugs_to_include ) && ! isset( $slugs_to_include[ $template_slug ] ) ) {
```

Instead of `array_search` use `isset` to check for the presence.

oandregal replied:
_Edit: sorry, accidentally merged this one. Just reverted it._

So, `$slugs_to_include` is going to be an array such as `array( 'some_slug', 'some_other_slug' )`. Using `isset` in this case will always return `false`, whether or not the key exists. Unless am I missing something?

mukeshpanchal27 replied:
```suggestion
			if ( ! empty( $slugs_to_include ) && ! in_array( $template_slug, $slugs_to_include, true ) ) {
```

Good catch.  In that case, `in_array` is more performant than `array_search`, so use it.

oandregal replied:
Happy to use `in_array` if that's preferred, though I doubt it has any performant impact in this scenario.

mukeshpanchal27 commented on src/wp-includes/block-template-utils.php:
```diff
@@ -317,6 +338,17 @@ function _get_block_templates_files( $template_type ) {
+			}
+
+			// Skip this item if its slug matches any of the slugs to skip.
+			if ( ! empty( $slugs_to_skip ) && ( false !== array_search( $template_slug, $slugs_to_skip, true ) ) ) {
```
```suggestion
			if ( ! empty( $slugs_to_skip ) && isset( $slugs_to_skip[ $template_slug ] ) ) {
```

Check similar comment above.

mukeshpanchal27 replied:
Check https://github.com/WordPress/wordpress-develop/pull/4097#discussion_r1150462689

oandregal replied:
https://github.com/WordPress/wordpress-develop/pull/4097/commits/d79f949dc839e87b02911298358346d048201d75

mukeshpanchal27 reviewed (commented):
Thanks @oandregal for the update. I take another round of review and left some review and questions.

In my test i found:
- The changes only affect block themes, so they did not improve performance for the classic themes. That performance comparison for classic themes is due to GHA variance.
- For block theme(Fresh setup with a post):
   - After the PR changes, the `_build_block_template_result_from_file` function called 7 times on the home page and 5 times on the post page, compared to 40 times on the home page and 22 times on the post page before the changes.

It's good improvement for the block theme.

oandregal committed:
Update src/wp-includes/block-template-utils.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

oandregal committed:
Update src/wp-includes/block-template-utils.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

oandregal committed:
Revert "Update src/wp-includes/block-template-utils.php"

This reverts commit 6f62ae250762c122ab779e0223558224ee401914.

oandregal committed:
Update src/wp-includes/block-template-utils.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

oandregal committed:
Use in_array instead of array_search

oandregal committed:
Make sure slugs are unique

mukeshpanchal27 reviewed (commented):
Thanks @oandregal for the changes. Can you share updated performance number?

oandregal commented:
> Thanks @oandregal for the changes. Can you share updated performance number?

I've rebased this upon `trunk` in preparation for commiting, and updated the issue description with the numbers reported by the perf CI job.

Commit:
Merge branch 'trunk' into improve/get-block-templates-performance

felixarntz commented:
@oandregal How did you get the numbers from the perf CI job? I'm asking since they only really run for `trunk` commits as far as I know. Did you trigger it yourself and look at the workflow run logs? The fact that classic themes see a notable improvement strikes me as potentially incorrect, since the change here should only affect block themes like you're saying.

felixarntz reviewed (approved):
@oandregal Outstanding work here! It's fair to say this is the most impressive single PR with major performance impact I have seen to date. 🥳 

Based on a [benchmark I just ran](https://docs.google.com/spreadsheets/d/1saGwDWET6ku-oP2BFfZzhTfwMuZZuzRTKxgyUM-i4-k/edit), this PR improves overall load time performance with block themes by an impressive ~15%.

Apologies for the delay in review, but it's safe to say this will be a major win for performance in WordPress 6.3 already - almost as good as all the improvements we got from WP 6.1 to 6.2 altogether.

mukeshpanchal27 reviewed (approved):
Thanks @oandregal, Awesome work.

oandregal commented:
> @oandregal How did you get the numbers from the perf CI job? I'm asking since they only really run for trunk commits as far as I know. Did you trigger it yourself and look at the workflow run logs?

:thinking: 

Oh. I looked at the [performance CI job that ran for this PR](https://github.com/WordPress/wordpress-develop/actions/runs/4768156334/jobs/8477156436?pr=4097). My understanding was that the numbers under "Print performance test results" was this PR and the numbers under "Print base line test results" was the baseline to compare to. Is that how should I look at them @felixarntz ?

The first time I ran my own numbers they were on the same range that the CI job, so I didn't bother doing my own again. Happy to if I'm reading this wrong.

> The fact that classic themes see a notable improvement strikes me as potentially incorrect, since the change here should only affect block themes like you're saying.

Yeah, I don't understand this either. I presumed it was related to the variability of the tests.

felixarntz commented:
@oandregal 
> Oh. I looked at the [performance CI job that ran for this PR](https://github.com/WordPress/wordpress-develop/actions/runs/4768156334/jobs/8477156436?pr=4097). My understanding was that the numbers under "Print performance test results" was this PR and the numbers under "Print base line test results" was the baseline to compare to. Is that how should I look at them @felixarntz ?

Ah that may be a misunderstanding and slightly confusing. The "base line test results" are not the current stats for `trunk`, but rather a re-run of getting metrics for `trunk` in the same baseline state that was originally measured (which I believe we have set to the last commit hash of WP 6.1.1). So this isn't the state of latest `trunk`.

Looking at the printed "performance test results" is a great idea though. What you need to compare that too though is the similar data for the last commit in `trunk` that is in your branch here.

All of that is of course not how this should work, but to be fair the performance CI workflow in its current stage was not yet meant to be used to assess PRs. Reporting diffs between the last commit in the branch and the PR automatically in a PR comment is high on the todo-list though :)

oandregal commented:
This has been committed at https://core.trac.wordpress.org/changeset/55687

oandregal commented:
This has created an issue with child themes, see instructions at https://github.com/WordPress/gutenberg/issues/53138 I'm looking into this.

oandregal commented:
Fix at https://github.com/WordPress/wordpress-develop/pull/4940

oandregal commented on src/wp-includes/block-template-utils.php:
```diff
@@ -957,34 +1019,14 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
-				continue;
-			}
-
-			$is_not_custom   = false === array_search(
```
This logic checks for both templates coming from the user ("custom" as it was defined here) and previously inserted templates (for example, it checks that a parent template is not added if the child already defined it). This wasn't properly ported and caused an issue.
